### DISCUSSION TITLE
Scanned data usage

### DISCUSSION THREAD

**Discussion by h3xty** [kel-z/HSR-Scanner#33] [General]
Not an issue, but got nowhere to ask. What are some sites/apps that you can input the scanned data into? I could find none, making the scanner kinda useless for me

**Comment by mattjneuber** [comment]
I was going to ask, as well, if there are any projects in development that will make use of this.  I've been thinking about writing something quick that will at least let me do a similar style of filtering that you can do in-game, but offline, but I don't have it in me to write out a whole equivalent to the Genshin optimizer.

**Comment by kel-z** [comment]
There aren't any apps that I know of yet that makes use of exporting and importing inventory data. If there is, you can let me know and I can try to make a conversion tool for it. 

That being said, I can quickly make an inventory viewer to visualize the scanned data. I'll see how far I can get with that this weekend. Feel free to make your own as well @mattjneuber

**Comment by kel-z** [comment]
Alright, this is what I managed to implement in two days as a proof of concept: [Star Rail Inventory Viewer](https://kel-z.github.io/starrail-dashboard/). It's purely a read-only app right now (can't edit items). If everything scanned correctly with HSR-Scanner, you'll see the exact stats that your characters have in-game.

The app also might blow up if you're importing an old JSON that's not from the latest HSR-Scanner release, since it doesn't really have any form of error-handling right now:))

When I have spare time, I'll implement the rest of the missing features (editing, more filtering options, character traces & eidolons, etc.)

**Comment by h3xty** [comment]
That is impressive! I love it so far, I hope you'll develop it further, looks really good so far

**Comment by VizualAbstract** [comment]
So, I just made something I'm calling a Relic Harmonizer... that will let you upload the JSON output. It's not as advanced as the Genshin Optimizer, but it's helped me get through all of my MoC stages.

[Relic Optimizer](https://coreycapetillo.com/honkai)

Just upload the file, select your character (it should show if the character has any currently-equipped relics), and click on "Harmonize Relics".

It'll generate a list of relics that are recommended based on your stat selections, and expose a little bit of the weights used to contribute to the recommendation.

It evaluates relic main stats, substats, and relic effects.

I'm sure it's buggy, and it takes 20-30 seconds or more to evaluate 1,000,000 combinations, but I should be able to improve on that in the next couple of days.

**Reply by kel-z** [reply]
Wow that is very promising! If you need a reference for the character traces part (i.e. what `stat_1`, `stat_2`…`stat_10` map to for each character), you can check out the [json](https://github.com/kel-z/HSR-Data/blob/cd8ce2f78ad4927e656e35f640b2f7e8ad1a701b/output/game_data_verbose.json#L13168) file that I use for my inventory viewer (though I haven't implemented the skill tree viewer part yet in the actual app, but the data is still there).

**Reply by VizualAbstract** [reply]
Thank you for that - I was trying to map the stat to the skill tree point so this could be calculated from the ([character skill tree values](https://github.com/Mar-7th/StarRailRes/blob/master/index_new/en/character_skill_trees.json#L2147-L2148)) and wrote this just to get the core idea implemented:

```
const StatToSkillTreePointMap = {
  stat_1: 'Point09',
  stat_2: 'Point12',
  stat_3: 'Point13',
  stat_4: 'Point10',
  stat_5: 'Point11',
  stat_6: 'Point15',
  stat_7: 'Point14',
  stat_8: 'Point16',
  stat_9: 'Point17',
  stat_10: 'Point18',
};
```

But had no idea if it was an accurate assumption 😩

**Reply by kel-z** [reply]
No problem. If you haven't noticed yet in the HSR-Scanner output, with respect to the data mine output, I got trace ordering for Erudition characters wrong for stat_3 + stat_4, and stat_6 + stat_7 (I didn't know of data mines back then, and of course I lose the 50/50 when deciding the correct Erudition tie-breaker order for those specific traces).

I [compensated for it (line 350)](https://github.com/kel-z/HSR-Data/blob/cd8ce2f78ad4927e656e35f640b2f7e8ad1a701b/src/extractors/game_data_verbose.py#L350) in my HSR-Data repo, but I'm thinking of correcting the actual output in the scanner and incrementing the version number in the JSON file to "3" so we can check and only accept the corrected version (so Erudition traces match the data mine outputs). What do you think?

**Reply by Azurmoo** [reply]
This scanner and website make a really good combination.  Definitely would like to see the harmonizer put some weight on the relic sets themselves, or allow you to select which sets to harmonize for a character.  Had a good laugh when it was suggesting four relics from four different sets.

**Reply by VizualAbstract** [reply]
@Azurmoo Yeah, I'd like to add the ability to do that. I'm hardcoding the weights at the moment, but I want to add an advanced feature that lets users set them with sliders. All in due time!

**Reply by kel-z** [reply]
Do you have a place for reporting bugs? I noticed that if you trying importing my [sample_output.json](https://github.com/kel-z/HSR-Scanner/blob/fd2a9a390c7682153855bcb80ab1530be6e8b89a/sample_output.json) file, Seele has 110.1% crit rate when it should be 66.3%.

**Comment by VizualAbstract** [comment]
> No problem. If you haven't noticed yet in the HSR-Scanner output, with respect to the data mine output, I got trace ordering for Erudition characters wrong for stat_3 + stat_4, and stat_6 + stat_7 (I didn't know of data mines back then, and of course I lose the 50/50 when deciding the correct Erudition tie-breaker order for those specific traces).
> 
> I [compensated for it (line 350)](https://github.com/kel-z/HSR-Data/blob/cd8ce2f78ad4927e656e35f640b2f7e8ad1a701b/src/extractors/game_data_verbose.py#L350) in my HSR-Data repo, but I'm thinking of correcting the actual output in the scanner and incrementing the version number in the JSON file to "3" so we can check and only accept the corrected version (so Erudition traces match the data mine outputs). What do you think?

I would of course love for correct data, but I'm just going to check for version number to handle minor cases like this

**Reply by kel-z** [reply]
Okay, yeah I agree it would save a lot headaches if I patched it now. I created a new release with this change just now.
